src/FSharp.AWS.DynamoDB/Diagnostics.fs

```diff
@@ -0,0 +1,70 @@
+namespace FSharp.AWS.DynamoDB
+
+module Diagnostics =
+    open System.Diagnostics
+    open Amazon.DynamoDBv2.Model
+    open System.Text.Json
+    
+    let private activitySource = new ActivitySource("FSharp.AWS.DynamoDB")
+
+    let hasListeners () : bool = activitySource.HasListeners()
+
+    let addActivityTags (tags: (string * obj) list) (activity: Activity) =
+        if notNull activity then
+            tags |> List.iter (fun (k, v) -> activity.AddTag(k, v) |> ignore)
+
+    let startTableActivity (tableName: string) (operation: string) (tags: (string * obj) list) : Activity =
+        let activity = activitySource.StartActivity(sprintf "%s %s" operation tableName, ActivityKind.Client)
+        activity
+        |> addActivityTags (
+            tags
+            |> List.append
+                [ ("db.system", "dynamodb")
+                  ("aws.dynamodb.table_names", [| tableName |])
+                  ("rpc.system", "aws-api")
+                  ("rpc.service", "DynamoDB")
+                  ("rpc.method", operation) ]
+        )
+        activity
+
+    let startMultipleTableActivity (tableNames: string seq) (operation: string) (tags: (string * obj) list) : Activity =
+        let activity = activitySource.StartActivity(operation, ActivityKind.Client)
+        activity
+        |> addActivityTags (
+            tags
+            |> List.append
+                [ ("db.system", "dynamodb")
+                  ("aws.dynamodb.table_names", tableNames |> Seq.toArray :> obj)
+                  ("rpc.system", "aws-api")
+                  ("rpc.service", "DynamoDB")
+                  ("rpc.method", operation) ]
+        )
+        activity
+
+    let addActivityCapacity (capacity: ConsumedCapacity seq) (activity: Activity) =
+        if notNull activity then
+            let value = capacity |> Seq.choose (function | null -> None | c -> Some (JsonSerializer.Serialize c)) |> Seq.toArray
```

Comment:
this fels wrong; surely there's a generic dictionary shape or something? or maybe it can be flattened into an entry per table, or... anything?!

Reply:
I don’t like it either, but the semantic conventions specifically define it like this. It would make more sense to me to use dynamic attribute keys - `aws.dynamodb.consumed_capacity.{index_name}.write_capacity` etc - other db traces use this for query params & the like. I can’t imagine many observability platforms go to the trouble of parsing JSON DynamoDB consumed capacity and doing anything useful with it.

I’m just as happy to leave it out, and maybe add metrics instead.

Reply:
sgtm - I think iterative tweaks make sense for this sort of work anyway - no subsittute for using it IRL to make you realise what works and doesnt